jetnet/utils/utils.py

```diff
@@ -148,3 +149,202 @@ def to_image(
+            else:
+                E = torch.sqrt(px ** 2 + py ** 2 + pz ** 2)
+            return torch.stack((E, px, py, pz), dim=-1)
+        return torch.stack((px, py, pz), dim=-1)
+
+    raise NotImplementedError(f'Current type {jet.type} not supported. Supported types are numpy.ndarray and torch.Tensor.')
+
+
+def get_polar_rel(
+    jet: np.ndarray or torch.Tensor,
+    input_cartesian: bool = False
+) -> np.ndarray or torch.Tensor:
+    """
+    Converts 3- or 4-vector features to relative polar coordinates.
+    Given jet feautures :math:`J = (J_\eta, J_\phi, J_{p_\mathrm{T}})` and particle features :math:`p = (\eta, \phi, p_\mathrm{T})`,
+    the relative coordinates are given by
+        .. math::
+            \eta_\mathrm{T}^\mathrm{rel} = \eta - J_\eta,\
+            \phi_\mathrm{T}^\mathrm{rel} = \phi - J_\phi,\
+            p_\mathrm{T}^\mathrm{rel} = p_\mathrm{T} / J_{p_\mathrm{T}}
+
+    Args:
+        jet (np.ndarray or torch.Tensor): array or tensor of a single jet of shape ``[num_particles, num_features]``
+          or multiple jets in cartesian or polar coordinates of shape ``[num_jets, num_particles, num_features]``.
+        input_cartesian (bool): Whether ``jet`` is in Cartesian coordinate. False if ``jet`` is in polar coordinates.
+          Default to False.
+
+    Returns:
+        type(jet): 2D (if inputted single jet) or 3D arrays/tensors of features in relative polar coordinates.
+
+    Raises:
+        NotImplementedError: If jet is not a numpy.ndarray or torch.Tensor.
+
+    """
+    if input_cartesian:
+        jet = get_polar(jet)  # Convert to polar first
+        if isinstance(jet, torch.Tensor):
+            jet_features = jet.sum(dim=-2)
+        elif isinstance(jet, np.ndarray):
+            jet_features = jet.sum(axis=-2)
+        else:
+            raise NotImplementedError(f'Current type {jet.type} not supported. Supported types are numpy.ndarray and torch.Tensor.')
+    else:
+        if isinstance(jet, torch.Tensor):
+            jet_features = get_polar(get_cartesian(jet).sum(dim=-2))
+        elif isinstance(jet, np.ndarray):
+            jet_features = get_polar(get_cartesian(jet).sum(axis=-2))
+        else:
+            raise NotImplementedError(f'Current type {jet.type} not supported. Supported types are numpy.ndarray and torch.Tensor.')
+
+    idx_eta, idx_phi, idx_pt = 0, 1, 2
+    pt = jet[..., idx_pt]
+    eta = jet[..., idx_eta]
+    phi = jet[..., idx_phi]
+
+    num_particles = jet.shape[-2]
+    if isinstance(jet, np.ndarray):
+        pt /= np.repeat(np.expand_dims(jet_features[..., idx_pt], axis=-1), num_particles, axis=-1)
+        eta -= np.repeat(np.expand_dims(jet_features[..., idx_eta], axis=-1), num_particles, axis=-1)
+        phi -= np.repeat(np.expand_dims(jet_features[..., idx_phi], axis=-1), num_particles, axis=-1)
+        return np.stack((eta, phi, pt), axis=-1)
+
+    pt /= jet_features[..., idx_pt].unsqueeze(dim=-1).repeat(1, num_particles)
```

Comment:
Even though they have the same effect, I'd turn this into an if elif block still with `elif isinstance(jet, torch.Tensor` just to be more clear.